warp10/src/main/java/io/warp10/continuum/gts/MetadataSelectorMatcher.java

```diff
@@ -0,0 +1,206 @@
+
+  private final Matcher classnamePattern;
+  private final Map<String, Matcher> labelsPatterns;
+  private final Map<String, Matcher> attributesPatterns;
+
+  private static final Pattern EXPR_RE = Pattern.compile("^(?<classname>[^{]+)\\{(?<labels>[^}]*)\\}(\\{(?<attributes>[^}]*)\\})?$");
+
+  public MetadataSelectorMatcher(String selector) throws WarpScriptException {
+
+    Matcher m = EXPR_RE.matcher(selector);
+    if (!m.matches()) {
+      throw new WarpScriptException("invalid syntax for selector.");
+    }
+
+    // read class selector
+    String classSelector = null;
+    try {
+      classSelector = WarpURLDecoder.decode(m.group("classname"), StandardCharsets.UTF_8);
+    } catch (UnsupportedEncodingException uee) {
+      // Can't happen, we're using UTF-8
+    }
+
+    // build class selector pattern
+    if ("~.*".equals(classSelector)) {
+      this.classnamePattern = null;
+    } else if ("=".equals(classSelector) || "~".equals(classSelector)) {
+      this.classnamePattern = Pattern.compile(Pattern.quote("")).matcher("");
+    } else if (classSelector.startsWith("~")) {
+      this.classnamePattern = Pattern.compile(classSelector.substring(1)).matcher("");
+    } else if (classSelector.startsWith("=")) {
+      this.classnamePattern = Pattern.compile(Pattern.quote(classSelector.substring(1))).matcher("");
+    } else {
+      this.classnamePattern = Pattern.compile(Pattern.quote(classSelector)).matcher("");
+    }
+
+    // read label selector
+    String labelsSelection = m.group("labels");
+    Map<String, String> labelsSelectors = null;
+    try {
+      labelsSelectors = GTSHelper.parseLabelsSelectors(labelsSelection);
+    } catch (ParseException pe) {
+      throw new WarpScriptException(pe);
+    }
+
+    if (0 == labelsSelectors.size()) {
+      this.labelsPatterns = null;
+    } else {
+      this.labelsPatterns = new HashMap<String, Matcher>(labelsSelectors.size());
+      // build label patterns map
+      for (Entry<String, String> l: labelsSelectors.entrySet()) {
+        if (l.getValue().startsWith("=")) {
+          this.labelsPatterns.put(l.getKey(), Pattern.compile(Pattern.quote(l.getValue().substring(1))).matcher(""));
+        } else { //starts with ~ , otherwise Parse Exception in GTSHelper.parseLabelsSelectors
+          this.labelsPatterns.put(l.getKey(), Pattern.compile(l.getValue().substring(1)).matcher(""));
+        }
+      }
+    }
+
+    // read attribute selector, if any
+    Map<String, String> attributesSelectors = null;
+    String attributesSelection = m.group("attributes");
+    if (attributesSelection != null) {
+      try {
+        attributesSelectors = GTSHelper.parseLabelsSelectors(attributesSelection);
+      } catch (ParseException pe) {
+        throw new WarpScriptException(pe);
+      }
+      this.attributesPatterns = new HashMap<String, Matcher>(attributesSelectors.size());
+      // build label patterns map
+      for (Entry<String, String> l: attributesSelectors.entrySet()) {
```

Comment:
Naming the variable 'a' instead of 'l' would be less disturbing since we are not iterating over (l)abels but (a)ttributes